### ISSUE TITLE
What is the scale for 3D handlandmark outputs?

### ISSUE THREAD

**Issue by pablovela5620** [google/mediapipe#156]
I have a few questions regarding the 3d tflite model outputs.
In the blog you explain that 
> we take Z-value from image depth map, if it exists per corresponding coordinate

The output of the 3d tflite model looks as follows

 P<sub>k</sub> = (u<sub>k</sub>,v<sub>k</sub>,z<sub>k</sub><sup>r</sup>)

where k is a specific landmark and z<sub>k</sub><sup>r</sup> is the root-relative depth

The [u,v] values look identical to the outputs from the 2d tflite model. My questions are regarding the z or depth coordinate

- What is the scale of the z coordinate i.e. is it centimeters, inches?
- What did the ground-truth value for the z coordinate look like?
- When stated that you take the z-value from a depth map does that mean that the training data consisted of rgb-d images that were manually annotated along with the synthetically generated data? 
- Would it be reasonable to say that the z predicted by the network for a specific landmark corresponds to the depth of that landmark relative to the root (wrist) landmark? i.e.  Z<sup>r</sup><sub>k</sub> = Z<sub>k</sub>  - Z<sub>root</sub> where Z<sub>k</sub> is the actual depth value of the k<sup>th</sup> landmark and  Z<sub>root</sub> is the actual depth value of the root (wrist) landmark and Z<sup>r</sup><sub>k</sub> is what is predicted by the 3d tflite network?
- Lastly is the depth value z<sub>k</sub><sup>r</sup> scale normalized? i.e. given P, P_hat is the scale normalized version where P_hat = C/s * P where s is the scale (Euclidian distance between a chosen joint and its parent) and C is a constant distance between specific pair of keypoints?

Thanks for all the great work you guys are doing, looking forward to your response!

**Comment by pablovela5620** [comment]
If there's anything I can do to get this question answered please let me know! If there's too many questions you may not answer that's fine too. If I could just know the scale of the root-relative z predictions (inches, millimeters, ..?) that would be the most helpful. I appreciate any response

**Comment by mcclanahoochie** [comment]
Hi
Some of the answers can be found in 
 https://github.com/google/mediapipe/issues/64 
 https://github.com/google/mediapipe/issues/99
 
I believe the z-coordinate is normalized relative to the wrist (close or far), and doesn't really have an absolute scale.

**Comment by pablovela5620** [comment]
Gotcha, so I don't actually need to know what the absolute scale is but rather the relative scale. So since it's root-relative I understand that the z-coordinates don't give me what the distance from the camera to the keypoint is, **but** I'm assuming that there is still a relative (normalized) scale that says what the distance is from the root (wrist) to the key point. For the training data there had to be SOME sort of scale or else it would be just arbitrary distances? If I could know what the relative (normalized) scale is then it would allow me to recover the absolute 3D coordinates using the camera intrinsics. Thank you again for the help

**Comment by mgyong** [comment]
@fanzhanggoogle Any comments on [@pablovela5620 comment](https://github.com/google/mediapipe/issues/156#issuecomment-545911833)?

**Comment by Sara533** [comment]
@mcclanahoochie 
It's obvious that he z-coordinate is normalized relative to the wrist.
And it doesn't really have an absolute scale.
What is the unit value of  the z-coordinate? 
Or there is no unit at all?

**Comment by mahdinobar** [comment]
> I have a few questions regarding the 3d tflite model outputs.
> In the blog you explain that
> 
> > we take Z-value from image depth map, if it exists per corresponding coordinate
> 
> The output of the 3d tflite model looks as follows
> 
> Pk = (uk,vk,zkr)
> 
> where k is a specific landmark and zkr is the root-relative depth
> 
> The [u,v] values look identical to the outputs from the 2d tflite model. My questions are regarding the z or depth coordinate
> 
>     * What is the scale of the z coordinate i.e. is it centimeters, inches?
> 
>     * What did the ground-truth value for the z coordinate look like?
> 
>     * When stated that you take the z-value from a depth map does that mean that the training data consisted of rgb-d images that were manually annotated along with the synthetically generated data?
> 
>     * Would it be reasonable to say that the z predicted by the network for a specific landmark corresponds to the depth of that landmark relative to the root (wrist) landmark? i.e.  Zrk = Zk  - Zroot where Zk is the actual depth value of the kth landmark and  Zroot is the actual depth value of the root (wrist) landmark and Zrk is what is predicted by the 3d tflite network?
> 
>     * Lastly is the depth value zkr scale normalized? i.e. given P, P_hat is the scale normalized version where P_hat = C/s * P where s is the scale (Euclidian distance between a chosen joint and its parent) and C is a constant distance between specific pair of keypoints?
> 
> 
> Thanks for all the great work you guys are doing, looking forward to your response!

Does someone have further answer to this comment?

**Comment by brianm-sra** [comment]
I am also wondering about this. Did you get any further answers?

**Comment by lbouis** [comment]
Any update on this? I am trying to understand how to get 3d coordinate from the hand landmarks, especially the z value is unclear to me.

**Comment by zhouwei5113** [comment]
> I have a few questions regarding the 3d tflite model outputs.
> In the blog you explain that
> 
> > we take Z-value from image depth map, if it exists per corresponding coordinate
> 
> The output of the 3d tflite model looks as follows
> 
> Pk = (uk,vk,zkr)
> 
> where k is a specific landmark and zkr is the root-relative depth
> 
> The [u,v] values look identical to the outputs from the 2d tflite model. My questions are regarding the z or depth coordinate
> 
> * What is the scale of the z coordinate i.e. is it centimeters, inches?
> * What did the ground-truth value for the z coordinate look like?
> * When stated that you take the z-value from a depth map does that mean that the training data consisted of rgb-d images that were manually annotated along with the synthetically generated data?
> * Would it be reasonable to say that the z predicted by the network for a specific landmark corresponds to the depth of that landmark relative to the root (wrist) landmark? i.e.  Zrk = Zk  - Zroot where Zk is the actual depth value of the kth landmark and  Zroot is the actual depth value of the root (wrist) landmark and Zrk is what is predicted by the 3d tflite network?
> * Lastly is the depth value zkr scale normalized? i.e. given P, P_hat is the scale normalized version where P_hat = C/s * P where s is the scale (Euclidian distance between a chosen joint and its parent) and C is a constant distance between specific pair of keypoints?
> 
> Thanks for all the great work you guys are doing, looking forward to your response!

Any further answer to this comment? I am also wondering how to use the z-value.

**Comment by mcclanahoochie** [comment]
please see https://github.com/google/mediapipe/issues/742#issuecomment-639104199

**Comment by matanster** [comment]
It is clear in the [current docs](https://google.github.io/mediapipe/solutions/hands#multi_hand_landmarks) now that Z is more or less the same scale as X. maybe it is scaled to the scale of X post-prediction by the pipeline. Haven't verified yet.
